pkg/scheduler/plugins/coescheduling/coescheduling.go

```diff
@@ -0,0 +1,215 @@
+	}
+	if subPrio1 != subPrio2 {
+		return subPrio1 > subPrio2
+	}
+
+	creationTime1 := cs.pgMgr.GetCreatTime(podInfo1)
+	creationTime2 := cs.pgMgr.GetCreatTime(podInfo2)
+	if creationTime1.Equal(creationTime2) {
+		return core.GetNamespaceSplicingName(podInfo1.Pod.Namespace, podInfo1.Pod.Name) < core.GetNamespaceSplicingName(podInfo2.Pod.Namespace, podInfo2.Pod.Name)
+	}
+	return creationTime1.Before(creationTime2)
+}
+
+// PreFilter
+// if non-strict-mode, we only do step1 and step2:
+// i.Check whether childes in Gang has met the requirements of minimum number under each Gang, and reject the pod if negative.
+// ii.Check whether the Gang has been timeout(check the pod's annotation,later introduced at Permit section) or is inited, and reject the pod if positive.
+// iii.Check whether the Gang has met the scheduleCycleValid check, and reject the pod if negative.
+// iv.Try update scheduleCycle, scheduleCycleValid, childrenScheduleRoundMap as mentioned above.
+func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, pod *v1.Pod) *framework.Status {
+	// If PreFilter fails, return framework.UnschedulableAndUnresolvable to avoid
+	// any preemption attempts.
+	if err := cs.pgMgr.PreFilter(ctx, pod); err != nil {
+		klog.Errorf("PreFilter failed pod: %v, err: %v ", core.GetNamespaceSplicingName(pod.Namespace, pod.Name), err)
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, err.Error())
+	}
+	return framework.NewStatus(framework.Success, "")
+}
+
+// PostFilter
+// i. If strict-mode, we will set scheduleCycleValid to false and release all assumed pods.
+// ii. If non-strict mode, we will do nothing.
+func (cs *Coscheduling) PostFilter(ctx context.Context, state *framework.CycleState, pod *v1.Pod,
+	filteredNodeStatusMap framework.NodeToStatusMap) (*framework.PostFilterResult, *framework.Status) {
+	return cs.pgMgr.PostFilter(ctx, pod, cs.frameworkHandler, Name)
+}
+
+// PreFilterExtensions returns a PreFilterExtensions interface if the plugin implements one.
+func (cs *Coscheduling) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Permit
+// we will calculate all Gangs in GangGroup whether the current number of assumed-pods in each Gang meets the Gang's minimum requirement.
+// and decide whether we should let the pod wait in Permit stage or let the whole gangGroup go binding
+func (cs *Coscheduling) Permit(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
+	waitTime, s := cs.pgMgr.Permit(ctx, pod)
+	var retStatus *framework.Status
+	switch s {
+	case core.PodGroupNotSpecified:
+		return framework.NewStatus(framework.Success, ""), 0
+	case core.PodGroupNotFound:
+		return framework.NewStatus(framework.Unschedulable, "Gang not found"), 0
+	case core.Wait:
+		klog.Infof("Pod: %v from gang: %v is waiting to be scheduled at Permit stage",
```

Comment:
I find some klog is gang is behind pod, some klog is reversed. please make the order same

Reply:
ok, all should be gangName first the podName